django/db/models/deletion.py

```diff
@@ -270,6 +270,43 @@ def delete(self):
         # number of objects deleted for each model label
         deleted_counter = Counter()
 
+        # caches query results of signals to avoid one hit per ['pre_', 'pos_']
+        pk_sets = dict()
+
+        def send_m2m_changed(model, obj, action):
+            """
+            Sends `m2m_changed` to all m2m relationships of this `model` that
+            contain at least one row in the table.
+            """
+            m2m_relations = [
+                f for f in model._meta.get_fields(include_hidden=True)
+                if f.many_to_many
+            ]
+
+            from django.db.models import ManyToManyField, ManyToManyRel
+            for m2m_relation in m2m_relations:
+                # for each relation, get all ids of the other side of
+                # the relation and send the signal with them
+                if isinstance(m2m_relation, ManyToManyField):
+                    attr_name = m2m_relation.name
+                    sender = m2m_relation.remote_field.through
+                else:
+                    assert(isinstance(m2m_relation, ManyToManyRel))
```

Comment:
What about using an `elif isinstance(m2m_relation, ManyToManyRel)` clause instead and issuing a `continue` in an `else` branch? This looks more safe in regard to third party fields that might exposed them as `many_to_many = True`.